File: trabalhoChatORB/TestedoChat3/Chat_Server.py
import sys
from omniORB import CORBA, PortableServer
import ChatApp__POA

class ChatServant(ChatApp__POA.Chat):
    def __init__(self):
        self.users = {}
        self.messages = []

    def sendMessage(self, username, message):
        formatted_message = f"{username}: {message}"
        print(formatted_message)
        self.messages.append(formatted_message)

    def receiveMessage(self):
        if self.messages:
            return self.messages.pop(0)
        else:
            return "No messages"

    def joinChat(self, username):
        self.users[username] = self._default_POA().servant_to_reference(self._this())
        print(f"{username} joined the chat.")

    def leaveChat(self, username):
        del self.users[username]
        print(f"{username} left the chat.")

    def listUsers(self):
        return ', '.join(self.users.keys())

    def sendPublicMessage(self, username, message):
        formatted_message = f"{username} (public): {message}"
        print(formatted_message)
        self.messages.append(formatted_message)

orb = CORBA.ORB_init(sys.argv, CORBA.ORB_ID)
poa = orb.resolve_initial_references("RootPOA")
poaManager = poa._get_the_POAManager()
poaManager.activate()

chatServant = ChatServant()
chatObject = chatServant._this()

print("Chat Server running...")
print("Object reference:", orb.object_to_string(chatObject))

orb.run()

File: trabalhoChatORB/C++/ChatServer.cpp
#include "Chat.hh"
#include <iostream>

class ChatServer_i : public POA_Chat::ChatServer {
public:
    void sendMessage(const char* username, const char* message) override {
        std::cout << username << ": " << message << std::endl;
    }

    char* receiveMessage() override {
        return CORBA::string_dup("Server received the message.");
    }
};

int main(int argc, char** argv) {
    try {
        CORBA::ORB_var orb = CORBA::ORB_init(argc, argv);

        PortableServer::POA_var poa = CORBA::RootPOA::_narrow(orb->resolve_initial_references("RootPOA"));

        PortableServer::POAManager_var pman = poa->the_POAManager();
        pman->activate();

        ChatServer_i* chatServer = new ChatServer_i();

        CORBA::Object_var obj = poa->servant_to_reference(chatServer);
        Chat::ChatServer_var chatServerRef = Chat::ChatServer::_narrow(obj);

        CORBA::String_var ior = orb->object_to_string(chatServerRef);
        std::cout << ior << std::endl;

        orb->run();
        delete chatServer;
        orb->destroy();
    } catch (const CORBA::Exception& e) {
        std::cerr << "Exception: " << e << std::endl;
    }

    return 0;
}

